exec/java-exec/src/main/java/org/apache/drill/exec/rpc/user/UserClient.java

```diff
@@ -140,8 +286,59 @@ protected Response handleReponse(ConnectionThrottle throttle, int rpcType, ByteB
 
   @Override
   protected void validateHandshake(BitToUserHandshake inbound) throws RpcException {
-//    logger.debug("Handling handshake from bit to user. {}", inbound);
-    if (inbound.getStatus() != HandshakeStatus.SUCCESS) {
+    switch (inbound.getStatus()) {
+    case SUCCESS:
+      break;
+    case AUTH_REQUIRED: {
+      logger.trace("Server requires authentication before proceeding.");
+      final String authMechanismToUse = AuthenticationUtil.getMechanismFromParams(connectionParams);
+      if (authMechanismToUse == null) {
+        throw new RpcException("Server requires authentication. But the client library could not deduce which " +
+            "mechanism to use from the given connection parameters.");
+      }
+
+      boolean isChosenMechanismSupported = false;
+      for (final String mechanism : inbound.getAuthenticationMechanismsList()) {
+        if (mechanism.equalsIgnoreCase(authMechanismToUse)) {
+          isChosenMechanismSupported = true;
+          break;
+        }
+      }
+      if (!isChosenMechanismSupported) {
+        throw new RpcException(
+            String.format("Server does not support the chosen mechanism: '%s' supported mechanisms: %s",
+                authMechanismToUse, inbound.getAuthenticationMechanismsList()));
+      }
+      logger.trace("Will try to authenticate using {} mechanism.", authMechanismToUse);
+
+      try {
+        switch (authMechanismToUse.toUpperCase()) {
+
+        case "PLAIN": {
+          final String name = connectionParams.getUserName();
+          final String password = connectionParams.getParam(ConnectionParams.PASSWORD);
+          saslClient = AuthenticationUtil.getPlainSaslClient(name, password);
+          break;
+        }
+
+        case "GSSAPI": {
+          final String principal = AuthenticationUtil.deriveKerberosName(connectionParams);
+          final String[] names = AuthenticationUtil.splitKerberosName(principal); // ignore names[2]
```

Comment:
Ignoring names[2] means we don't care user domain. Are we assuming server and client  are in the same domain?